tsl/src/fdw/option.c

```diff
@@ -273,3 +283,58 @@ option_get_from_options_list_int(List *options, const char *optionname, int *val
 
 	return found;
 }
+
+List *
+option_extract_join_ref_table_list(const char *table_string)
+{
+	List *ref_table_oids = NIL;
+	List *ref_table_list;
+	ListCell *lc;
+
+	/* SplitIdentifierString scribbles on its input, so pstrdup first */
+	if (!SplitIdentifierString(pstrdup(table_string), ',', &ref_table_list))
+	{
+		/* syntax error in name list */
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("parameter \"%s\" must be a list of reference table names",
+						"join_reference_tables")));
+	}
+
+	foreach (lc, ref_table_list)
+	{
+		char *tablename = (char *) lfirst(lc);
+
+		RangeVar *rangevar = makeRangeVarFromNameList(stringToQualifiedNameList(tablename));
+
+		Oid relOid = RangeVarGetRelidExtended(rangevar,
+											  AccessShareLock,
+											  RVR_MISSING_OK,
+											  NULL /* callback */,
+											  NULL /* callback args*/);
+
+		if (!OidIsValid(relOid))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+					 errmsg("table \"%s\" does not exist.", tablename)));
+		}
+
+		/* Validate the relation type */
+		Relation rel = table_open(relOid, NoLock);
+
+		if (rel->rd_rel->relkind != RELKIND_RELATION)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+					 errmsg("table \"%s\" is not a relation. Only relations can be used as "
```

Comment:
The message should be the other way around? "relation is not an ordinary table".

Reply:
@akuzm Thanks, it is changed.